From c294749f8751cb8581edb1b832ac27b5d14de02b Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Fri, 29 Oct 2010 21:00:54 +0100
Patch-mainline: 2.6.38
References: fate#310916
Git-commit: 4066c0ae13a8388b9f2a29cad60da330b578eff2
Subject: [PATCH 1825/2588] drm/i915/debugfs: Display the contents of the BLT
 and BSD status pages

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
(cherry picked from commit 4066c0ae13a8388b9f2a29cad60da330b578eff2)

Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 drivers/gpu/drm/i915/i915_debugfs.c |   16 +++++++++++++---
 1 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_debugfs.c b/drivers/gpu/drm/i915/i915_debugfs.c
index f725f1f..5af2114 100644
--- a/drivers/gpu/drm/i915/i915_debugfs.c
+++ b/drivers/gpu/drm/i915/i915_debugfs.c
@@ -445,10 +445,18 @@ static int i915_hws_info(struct seq_file *m, void *data)
 	struct drm_info_node *node = (struct drm_info_node *) m->private;
 	struct drm_device *dev = node->minor->dev;
 	drm_i915_private_t *dev_priv = dev->dev_private;
-	int i;
+	struct intel_ring_buffer *ring;
 	volatile u32 *hws;
+	int i;
+
+	switch ((uintptr_t)node->info_ent->data) {
+	case RENDER_RING: ring = &dev_priv->render_ring; break;
+	case BSD_RING: ring = &dev_priv->bsd_ring; break;
+	case BLT_RING: ring = &dev_priv->blt_ring; break;
+	default: return -EINVAL;
+	}
 
-	hws = (volatile u32 *)dev_priv->render_ring.status_page.page_addr;
+	hws = (volatile u32 *)ring->status_page.page_addr;
 	if (hws == NULL)
 		return 0;
 
@@ -1086,7 +1094,9 @@ static struct drm_info_list i915_debugfs_list[] = {
 	{"i915_gem_seqno", i915_gem_seqno_info, 0},
 	{"i915_gem_fence_regs", i915_gem_fence_regs_info, 0},
 	{"i915_gem_interrupt", i915_interrupt_info, 0},
-	{"i915_gem_hws", i915_hws_info, 0},
+	{"i915_gem_hws", i915_hws_info, 0, (void *)RENDER_RING},
+	{"i915_gem_hws_blt", i915_hws_info, 0, (void *)BLT_RING},
+	{"i915_gem_hws_bsd", i915_hws_info, 0, (void *)BSD_RING},
 	{"i915_ringbuffer_data", i915_ringbuffer_data, 0, (void *)RENDER_RING},
 	{"i915_ringbuffer_info", i915_ringbuffer_info, 0, (void *)RENDER_RING},
 	{"i915_bsd_ringbuffer_data", i915_ringbuffer_data, 0, (void *)BSD_RING},
-- 
1.7.6

